devolutions-gateway/src/api/network_scan.rs

```diff
@@ -0,0 +1,123 @@
+use std::net::IpAddr;
+
+use axum::{
+    extract::{ws::Message, State, WebSocketUpgrade},
+    response::Response,
+};
+
+use network_scanner::scanner::NetworkScannerParams;
+use serde::Serialize;
+
+use crate::{http::HttpError, DgwState};
+
+pub async fn handler(
+    _: State<DgwState>,
+    ws: WebSocketUpgrade,
+    query_params: axum::extract::Query<NetworkScanQueryParams>,
+) -> Result<Response, HttpError> {
+    let scanner_params: NetworkScannerParams = query_params.0.into();
+
+    let scanner = network_scanner::scanner::NetworkScanner::new(scanner_params).map_err(|e| {
+        tracing::error!("Failed to create network scanner: {:?}", e);
+        HttpError::internal().build(e)
+    })?;
+
+    let res = ws.on_upgrade(move |mut websocket| async move {
+        let stream = scanner.start().expect("Failed to start network scanner");
```

Comment:
issue: `expect` will panics on failure. If you can’t return an error, log an error instead.

Reply:
updated